public/components/trace_analytics/components/common/helper_functions.tsx

```diff
@@ -677,3 +675,83 @@ export const parseHits = (payloadData: string): ParsedHit[] => {
+
+export const InjectElementsIntoGrid = (
+  rowCount: number,
+  maxDisplayRows: number,
+  tracesTableMode: string,
+  loadMoreHandler?: () => void
+) => {
+  useEffect(() => {
+    setTimeout(() => {
+      const toolbar = document.querySelector('.euiDataGrid__controls') as HTMLElement;
+      if (toolbar && rowCount > maxDisplayRows) {
+        toolbar.style.display = 'flex';
+        toolbar.style.alignItems = 'center';
+        toolbar.style.justifyContent = 'space-between';
+
+        let warningDiv = toolbar.querySelector('.trace-table-warning') as HTMLElement;
+        if (!warningDiv) {
+          warningDiv = document.createElement('div');
+          warningDiv.className = 'trace-table-warning';
+          warningDiv.style.marginLeft = 'auto';
+          warningDiv.style.color = '#a87900';
+          warningDiv.style.fontSize = '13px';
+          warningDiv.style.display = 'flex';
+          warningDiv.style.alignItems = 'center';
+          warningDiv.style.padding = '2px 8px';
+          warningDiv.style.borderRadius = '4px';
+          warningDiv.style.whiteSpace = 'nowrap';
+
+          const icon = document.createElement('span');
+          icon.className = 'euiIcon euiIcon--medium euiIcon--warning';
+          icon.setAttribute('aria-label', 'Warning');
+          icon.style.marginRight = '4px';
+
+          const strongElement = document.createElement('strong');
+          strongElement.textContent = `${maxDisplayRows}`;
+
+          const textSpan = document.createElement('span');
+          textSpan.appendChild(strongElement);
+          textSpan.appendChild(document.createTextNode(' results shown out of '));
+          textSpan.appendChild(document.createTextNode(` ${rowCount}`));
+
+          warningDiv.appendChild(icon);
+          warningDiv.appendChild(textSpan);
+
+          toolbar.appendChild(warningDiv);
+        }
+      }
+
+      const pagination = document.querySelector('.euiDataGrid__pagination') as HTMLElement;
+
+      if (tracesTableMode !== 'traces') {
+        if (pagination) {
+          const existingLoadMoreButton = pagination.querySelector('.trace-table-load-more');
+          if (existingLoadMoreButton) {
+            existingLoadMoreButton.remove();
+          }
+        }
+        return;
+      }
+
+      if (pagination && loadMoreHandler) {
+        pagination.style.display = 'flex';
+        pagination.style.alignItems = 'center';
+        pagination.style.justifyContent = 'space-between';
+
+        let loadMoreButton = pagination.querySelector('.trace-table-load-more') as HTMLElement;
+        if (!loadMoreButton) {
+          loadMoreButton = document.createElement('button');
+          loadMoreButton.className = 'trace-table-load-more euiButtonEmpty euiButtonEmpty--text';
+          loadMoreButton.style.marginLeft = '12px';
+          loadMoreButton.innerText = 'Load more data';
+
+          loadMoreButton.onclick = () => loadMoreHandler();
+
+          pagination.appendChild(loadMoreButton);
```

Comment:
DOM manipulation is hard to maintain and might break when OUI library changes. is it possible to use existing data grid functions to do this? for example https://oui.opensearch.org/1.19/#/tabular-content/data-grid-styling-and-control#additional-controls-in-the-toolbar

Reply:
Unfortunately, Data grid “Additional controls” only allows insertion to the left of the default controls. We are currently using that to add the Span/Trace selection and Fullscreen buttons. The component does not allow for buttons to be split into different locations to have the warning in the right corner. It also doesn’t allow for manipulation of the pagination to insert the “Load more” button at the bottom right. With current UX requirements, I think DOM manipulation is the best solution; pending  a refactor / adjustments of the OUI component.